### PR TITLE
Add extension context to SessionState.

### LINKED ISSUE #8926: How to provide a custom `Context` in `TableProvider`'s `scan`?
### Is your feature request related to a problem or challenge?

`TableProvider`'s `scan` often load a lot of data from some place. if somehow this request is cancelled, we can drop the `scan` .
can save use a lot of cpu times. right now there is no way to pass the custom context for each `scan` .

### Describe the solution you'd like

since scan take a `SessionState` as parameter . 
we can add a field to this struct.
~~~
pub struct SessionState {
   ...
   context : Option<Arc<dyn Any>>,
   ...
}
~~~

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by yuyang-ok**
This trying to fix #8926.

### CODE REVIEW

File: datafusion/core/src/execution/context/mod.rs [hunk 1]
```diff
@@ -1869,6 +1872,18 @@ impl SessionState {
     pub fn version(&self) -> &str {
         env!("CARGO_PKG_VERSION")
     }
+
+    /// Return extension context.
+    pub fn with_extension_context(&self) -> Option<Arc<dyn Any + Send + Sync>> {
```
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:1877]: I think `with` prefixed functions are normally used to set fields in builder style code. I think accessors would be more normally named like this:

```suggestion
    pub fn extension_context(&self) -> Option<Arc<dyn Any + Send + Sync>> {
```

File: datafusion/core/src/execution/context/mod.rs [hunk 2]
```diff
@@ -1869,6 +1872,18 @@ impl SessionState {
...
+        self.extension_context.clone()
+    }
+
+    /// Return extension context mut.
+    pub fn with_extension_context_mut(
```
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:1882]:
```suggestion
    pub fn extension_context_mut(
```

**Review by alamb — COMMENTED** [review]
Thank you @yuyang-ok  -- this is looking pretty good to me. 

I left some naming suggestions

In order to merge this PR, I think we need to add a test showing how to use this feature (to ensure everything needed is indeed connected and to make sure we don't accidentally break the feature with some future refactor).

This can either take the form of an end to end test in https://github.com/apache/arrow-datafusion/tree/main/datafusion/core/tests/user_defined or  an example in https://github.com/apache/arrow-datafusion/tree/main/datafusion-examples/examples

Does that make sense? Is it something you are able to try?

**Comment by alamb** [comment]
I don't think this PR is waiting on feedback, so marking it as a draft. Please mark it as ready for review when it is ready for another look 🙏

**Comment by yuyang-ok** [comment]
@alamb  sorry.

**Comment by alamb** [comment]
> @alamb sorry.

Nothing to be sorry about at all -- I am just trying to make sure all PRs get a timely review and it is hard to keep track of which ones have and have not had reviews yet 

Is there something I can do to help this PR along? or maybe @yliang412 might have some thoughts?

**Comment by alamb** [comment]
Any update @yuyang-ok ?

**Comment by yuyang-ok** [comment]
sorry @alamb  these days are Chinese new year.  I think maybe close this PR for a while.
I will fix this later.

**Pull request closed without merging**